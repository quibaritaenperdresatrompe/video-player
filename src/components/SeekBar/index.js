import glamorous from 'glamorous'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

const SeekBarContainer = glamorous.div(({ size }) => ({
  margin: '0.5em 0',
  backgroundColor: 'hsla(0, 0%, 100%, 0.8)',
  height: size,
  opacity: '0.8',
  position: 'relative',
  ':hover': {
    opacity: '1',
  },
}))

const StyledSeekBar = glamorous.input(({ size }) => {
  const thumbStyle = {
    appearance: 'none',
    backgroundColor: 'hsl(0, 68%, 50%)',
    border: 0,
    borderRadius: '50%',
    height: `calc(${size} * 5)`,
    width: `calc(${size} * 5)`,
  }

  return {
    appearance: 'none',
    backgroundColor: 'transparent',
    border: 0,
    cursor: 'pointer',
    height: size,
    margin: 0,
    outline: 'none',
    position: 'relative',
    width: '100%',
    '::-webkit-slider-thumb': thumbStyle,
    '::-moz-range-thumb': thumbStyle,
    '::-ms-thumb': thumbStyle,
  }
})

const StyledProgressBar = glamorous.div(({ progress }) => ({
  backgroundColor: 'hsl(0, 68%, 50%)',
  height: '100%',
  left: 0,
  position: 'absolute',
  top: 0,
  width: progress === 100
    ? `${progress}%`
    : `calc(${progress}% + 1px)`,
}))

class SeekBar extends Component {
  constructor() {
    super()
    this.state = {
      isSeeking: false,
    }
  }

  handleChange = ({ target: { value } }) => {
    this.props.seekTo(parseFloat(value, 10))
    this.setState(() => ({
      isSeeking: true,
    }))
  }

  render() {
    const {
      currentTime,
      duration,
      size,
    } = this.props

    const progress = duration === 0
      ? 0
      : (currentTime * 100) / duration

    return (
      <SeekBarContainer size={size}>
        <StyledSeekBar
          max={duration}
          onChange={this.handleChange}
          size={size}
          step='any'
          type='range'
          value={currentTime}
        />
        <StyledProgressBar
          progress={progress}
          size={size}
        />
      </SeekBarContainer>
    )
  }
}

SeekBar.propTypes = {
  currentTime: PropTypes.number,
  duration: PropTypes.number,
  seekTo: PropTypes.func,
  size: PropTypes.string,
}

SeekBar.defaultProps = {
  currentTime: 0,
  duration: 0,
  seekTo: () => undefined,
  size: '0.2em',
}

export default SeekBar
