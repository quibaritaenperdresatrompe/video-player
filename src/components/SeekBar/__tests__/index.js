import SeekBar from '../'

import itRendersAllMutations from '../../../utils/jest-it-renders-all-mutations'

const mutations = [
  {
    name: 'without props',
    props: null,
  },
  {
    name: 'without currentTime and duration',
    props: {
      currentTime: 9,
      duration: 10,
    },
  },
  {
    name: 'when it is at the end',
    props: {
      currentTime: 10,
      duration: 10,
    },
  },
]

describe(SeekBar.name, () => {
  itRendersAllMutations(SeekBar, mutations)

  describe('handlers', function scope() {
    beforeEach(() => {
      this.Player = new SeekBar()
    })

    describe('handleChange', () => {
      test('it calls props.seekTo and setState', () => {
        this.Player.props = {
          seekTo: jest.fn(),
        }
        this.Player.setState = jest.fn()
        const event = { target: { value: 99 } }
        this.Player.handleChange(event)
        expect(this.Player.props.seekTo).toHaveBeenCalledWith(event.target.value)
        expect(this.Player.setState).toHaveBeenCalled()
      })
    })
  })
})
