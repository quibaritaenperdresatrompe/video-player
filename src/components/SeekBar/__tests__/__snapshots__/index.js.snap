// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeekBar when it is at the end it renders correctly 1`] = `
<div
  className="css-1tbhals"
>
  <input
    className="css-19h4nxv"
    max={10}
    onChange={[Function]}
    size="0.2em"
    step="any"
    type="range"
    value={10}
  />
  <div
    className="css-z5eujh"
  />
</div>
`;

exports[`SeekBar without currentTime and duration it renders correctly 1`] = `
<div
  className="css-1tbhals"
>
  <input
    className="css-19h4nxv"
    max={10}
    onChange={[Function]}
    size="0.2em"
    step="any"
    type="range"
    value={9}
  />
  <div
    className="css-pvxstt"
  />
</div>
`;

exports[`SeekBar without props it renders correctly 1`] = `
<div
  className="css-1tbhals"
>
  <input
    className="css-19h4nxv"
    max={0}
    onChange={[Function]}
    size="0.2em"
    step="any"
    type="range"
    value={0}
  />
  <div
    className="css-1voc2x1"
  />
</div>
`;
